File: languages/scala/src/model/BenchmarkOutput.scala
package model

import model.TimeFormat.TimeFormat

/**
  * Interface para captura de estados de tempo (benchmark)
  */
trait BenchmarkOutput {

  /**
    * Inicio da captura de estado
    * @param tag Nome da captura referente
    */
  def start(tag: String)

  /**
    * Fim da captura de estado
    * @param tag Nome da captura referente
    */
  def end(tag: String)

  /**
    * Resultado da captura de estado especifica
    * @param tag Nome da captura referente
    * @param format Formato do resultado
    * @return Tempo decorrido entre o inicio e o fim da captura de estado
    */
  def result(tag: String, format: TimeFormat): Double

  /**
    * Resultado de todas as capturas de estado
    * @param format Formato do resultado
    * @return Mapa contendo o tempo decorrido entre o inicio e o fim da captura de estado para cada estado gerado.
    */
  def result(format: TimeFormat): Map[String, Double]

  /**
    * Exporta o resultado no formato de um arquivo
    * @param fileName Nome do arquivo de saida
    * @param format Formato do resultado
    */
  def export(fileName: String, format: TimeFormat)
}

File: languages/dotnet/src/model/BenchmarkOutput.cs
using System;
using System.Collections.Generic;

namespace experimento.src.model {
 
    /// <summary>
    /// Interface para captura de estados de tempo (benchmark)
    /// </summary>
    public interface BenchmarkOutput {      
 
        /// <summary>
        /// Inicio da captura de estado
        /// </summary>
        /// <param name="tag">Nome da captura referente</param>
        void start(string tag);

        /// <summary>
        /// Fim da captura de estado
        /// </summary>
        /// <param name="tag">Nome da captura referente</param>
        void end(string tag);

        /// <summary>
        /// Resultado da captura de estado especifica
        /// </summary>
        /// <param name="tag">Nome da captura referente</param>
        /// <param name="format">Formato do resultado</param>
        /// <returns>Tempo decorrido entre o inicio e o fim da captura de estado</returns>
        double result(string tag, TimeFormat format);

        /// <summary>
        /// Resultado de todas as capturas de estado
        /// </summary>
        /// <param name="format">Formato do resultado</param>
        /// <returns>Mapa contendo o tempo decorrido entre o inicio e o fim da captura de estado para cada estado gerado.</returns>
        Dictionary<string, double> result(TimeFormat format);

        /// <summary>
        /// Exporta o resultado no formato de um arquivo
        /// </summary>
        /// <param name="fileName">Nome do arquivo de saida</param>
        /// <param name="format">Formato do resultado</param>
        void export(string fileName, TimeFormat format);
    }
}